Observing Randomness blogpost (#139)
drand/website: docs/blog/2022-08-12-observing-randomness.md

## What is drand?

[Drand](https://drand.love/) is a public, verifiable, unpredictable and unbiased randomness beacon. It is based on the idea that, to generate randomness in a reliable and trustworthy way, you cannot depend on a single entity to control the infrastructure where it runs. For this, we have sought the collaboration of multiple organizations to run their own drand daemons, forming the [League of Entropy](https://blog.cloudflare.com/league-of-entropy/) (*LoE*) - a distributed organization where each of the participating partners runs a drand node. There are other resources, such as the project’s [documentation pages](https://drand.love/docs/), which explain the cryptographic details regarding how randomness is generated by multiple parties using threshold cryptography. This blog post will instead concentrate on the challenges we faced when we tried to design, implement and deploy monitoring infrastructure for this distributed network.

> As a little bit of background, a drand *network* is composed of multiple *nodes*. Each network generates one (or more) *beacons*, each of which generates separate randomness beacons at a specific interval, called a *round*. The number of nodes associated with each beacon is fixed at any given point in time and is only changed during a *ceremony*. During the ceremony a *threshold* is set, which is the minimum number of nodes that need to be connected to the network at any given time for randomness generation to succeed.

Reviewer: @AnomalRoil how critical would you say that this change is? I was going for a higher-level description, going from the network down to the nodes instead of from the nodes up to the network.

> As a little bit of background, a drand *network* is composed of multiple *nodes*. Each network generates one (or more) *beacons*, each of which generates separate randomness beacons at a specific interval, called a *round*. The number of nodes associated with each beacon is fixed at any given point in time and is only changed during a *ceremony*. During the ceremony a *threshold* is set, which is the minimum number of nodes that need to be connected to the network at any given time for randomness generation to succeed.

Reviewer 2: Well my main issue is that IMO "Each network generates one (or more) *beacons*, each of which generates separate randomness beacons" is wrong. 

A network doesn't generate beacons, or it does so if we mean the rounds themselves.
And saying that a beacon generates a beacon seems weird.

Maybe instead?
```suggestion
As a little bit of background, a drand *network* is composed of multiple *nodes*. A given network is also sometimes called a *beacon chain*, since each network generates separate randomness beacons at a specific interval, called also a *round*. The number of nodes on a network is fixed at any given point in time and is only changed during a *ceremony*. During the ceremony a *threshold* is set, which is the minimum number of nodes that need to be connected to the network at any given time for randomness generation to succeed.
```

## The need for monitoring

To ensure security, we do not allow just any host to join the network at any time. Whenever a host joins or leaves the network, we hold what is called a “ceremony”, where cryptographic keys are exchanged. In between ceremonies the list of hosts that participate in a drand network is static. Ceremonies are usually held quarterly, regardless of whether any nodes join or leave the network. Before each ceremony, an allowlist is distributed to all the participants, which contains the IP addresses of all nodes in the network. This is done so that LoE members can open the appropriate firewall ports.

> One issue with the ceremonies is that it is a synchronous operation where all node operators need to be at the keyboard at the same (window of) time and execute certain commands in their nodes at the correct moment and in the correct order. Since node operators are based around the world, it is a challenge to coordinate a time that is convenient for everyone, and that does not even account for unforeseen reasons of why a node operator might be unavailable. In the past this has led to delaying or even postponing some ceremonies.

Reviewer: The issue is not really with the nodes, but with the operators. An operator might have nodes in 5 different countries (such as PL with nodes in multiple AWS regions) but the important part is that those of us who operate drand are mostly located in Europe.

> One issue with the ceremonies is that it is a synchronous operation where all node operators need to be at the keyboard at the same (window of) time and execute certain commands in their nodes at the correct moment and in the correct order. Since node operators are based around the world, it is a challenge to coordinate a time that is convenient for everyone, and that does not even account for unforeseen reasons of why a node operator might be unavailable. In the past this has led to delaying or even postponing some ceremonies.

Reviewer 3: `for unforeseen reasons of why a node operator might be unavailable` -of

> One issue with the ceremonies is that it is a synchronous operation where all node operators need to be at the keyboard at the same (window of) time and execute certain commands in their nodes at the correct moment and in the correct order. Since node operators are based around the world, it is a challenge to coordinate a time that is convenient for everyone, and that does not even account for unforeseen reasons of why a node operator might be unavailable. In the past this has led to delaying or even postponing some ceremonies.

Reviewer 2: My goal was to recall that drand nodes are located all over the world

> There are other things that can go wrong during a ceremony. One of the nodes might not be running the appropriate version of `drand`. The node operator might restart their node in the middle of the ceremony, which then puts their node in a wrong state to participate in a ceremony. There might be connectivity problems which do not allow a node to correctly participate. And so on and so forth. Up until now, the coordination between operators has been through off-band channels (such as email or Slack), which resulted in postponed or cancelled ceremonies. The difficulties in coordinating a ceremony through such means surfaced the need to have full visibility into the drand network and the state of each node participating in it, in order to ensure that all nodes are in the correct state and that everything happens at the right time and in the right order.

Reviewer 3: The reasons might be clearer in a bullet point list

Reviewer: I thought of putting it in text since it's not an exhaustive list, I think that it reads better this way.

## Monitoring architecture requirements

To properly monitor the drand network we have some requirements:

**Allow any LoE member to create their own monitoring infrastructure**

We did not want to mandate using a specific metrics aggregation platform (e.g. Prometheus or InfluxDB), instead allowing for each organization to select the one that best meets their needs or with which they are most familiar.

**Ease of deployment**

We did not want to burden operators with complex deployment strategies, so we decided to modify the drand binary itself instead of adding a new binary to export the metrics.

**Minimize the need for opening network ports**

> Since drand nodes are run by multiple organizations on their own networks, any changes to the firewall configuration increases friction and reduces the probability of success. Ideally no network ports other than the one used for generating randomness should need to be opened. At the very least no more incoming ports should be needed.

Reviewer: I think `ports` (plural) is more correct here

> Since drand nodes are run by multiple organizations on their own networks, any changes to the firewall configuration increases friction and reduces the probability of success. Ideally no network ports other than the one used for generating randomness should need to be opened. At the very least no more incoming ports should be needed.

Reviewer 3: both are fine here imo

**Use industry-standard formats where possible**

To decrease the ramp-up time for integrating into an organization’s monitoring system, we need to use an industry-standard format to publish metrics.

**Ability to see the state of the whole network from any point in the network**

> To allow all LoE members to create their own dashboards and alerts, we need to enable each node to provide a full view of the whole network.

Reviewer 2: ```suggestion
To allow all LoE members to create their own dashboards and alerts, we need to allow each node to provide a full view into the whole network.
```

Reviewer 3: view of the whole network sounds better imo

Reviewer 3: but does the node _provide_ a full view or _receive_ a full view?

Reviewer: you can get a full view from any node, so the node provides a full view (for which it receives data from each of the other nodes)

Reviewer: I like `enable` rather than `allow`, since it's not a permissions problem but rather a functionality problem.

**(Close to) real-time view of the network state**

Especially during ceremonies, it is important to see changes in node state as soon as possible (< 1 minute) after they happen.

## Design decisions

The above requirements led us to the following design:

**Each node exposes its own metrics in Prometheus format**

Prometheus is an industry-standard format, and each node exposing its own metrics allows anyone to monitor their own node directly.

**Additionally, each node is able to retrieve any other connected node’s metrics and expose them in Prometheus format**

This enables not only Protocol Labs but any LoE member to create their own dashboards which show the state of the whole network.

**The transport of metrics between nodes is done using the same mechanism that drand uses for inter-node communication when generating randomness (GRPC)**

> With this we allow metrics to be sent between nodes, while not requiring opening any additional network ports.

Reviewer 2: ```suggestion
With this we allow metrics to be sent between nodes, without requiring to open any additional network port.
```

Reviewer 3: disagree with both 'to open' and 'port'

Reviewer 2: You're not opening any, so why plural?

Reviewer 3: I don't know why, it's just right ;p

Reviewer: Plural "feels" more right. I also disagree with the change, `opening` is correct.

**For the Protocol Labs dashboard, gather the metrics using Telegraf and export them to InfluxDB, which we can access via Grafana Cloud**

This enables us to gather metrics from the whole swarm into a central location and create an easy-to-use dashboard. The scraping interval is set to one minute, and it can be lowered during ceremonies in order to get faster network state updates.

## Metrics design

To start with, we had to decide what we needed to monitor on each node. We came up with the following minimal information requirements for running a smooth ceremony:

- Version of drand that the node is running, in order to ensure that all nodes are running the same/correct version.

> - Whether nodes have restarted their daemon ahead of the ceremony, to ensure that the certificate is still valid and to clear the process’ memory and storage.

Reviewer 2: ```suggestion
- When nodes have last restarted their daemon, to ensure a given node wasn't restarted during the ceremony.
```

Reviewer: Here we're talking about the initial requirements, which was some indication that the daemon had been restarted before the ceremony. The *solution* was to show the time of the last restart.

- Whether nodes have connectivity to all other nodes.
- When and if the node started the ceremony process.

> - What is the current state of the node related to the ceremony.

Reviewer 3: `what the current state of the node is with relation to the ceremony` or perhaps just `the current ceremony status of the node`

Reviewer 3: Albeit is there a difference between `When and if the node started the ceremony process` and `What is the current state of the node related to the ceremony`? They seem to be overlapping

We also came up with the idea of using the dashboard for ongoing monitoring of the network (not only during ceremonies), for which we added the following for each beacon in each node:

- When was the last randomness generated.

> - Which is the last round generated by the node.

Reviewer 2: ```suggestion
- What is the last round generated by the node for that beacon chain.
```

- What is the time difference between this node and other nodes (including network round-trip times).

> The drand code was already instrumented to provide some of the needed metrics via a Prometheus HTTP endpoint, while others needed to be added.

Reviewer 2: ```suggestion
The drand code was already instrumented to provide some of the required metrics via a Prometheus HTTP endpoint, while others needed to be added.
```

Reviewer 2: To avoid the "needed" repetition.

## Inter-node communications

> As mentioned above, one of the requirements was to be able to view the state of the whole drand network without requiring node operators to open any additional holes in their firewall, nor to connect to any hosts other than those included in the drand network itself. Therefore, it was necessary to enable the possibility of fetching a remote node’s metrics via the same method used to coordinate randomness generation: that is, through GRPC.

Reviewer 2: ```suggestion
As mentioned above, one of the requirements was to be able to view the state of the whole drand network without requiring node operators to open any additional holes in their firewall, nor to connect to any host other than those included in the drand network itself. Therefore, it was necessary to enable the possibility of fetching a remote node’s metrics via the same method used to coordinate randomness generation: that is, through GRPC.
```

Reviewer: Same as with `ports` vs `port` elsewhere, to me the plural feels "more correct".

> We also wanted to reduce the amount of metrics traffic between nodes, so we went with a pull system where any node can pull any other node’s metrics when requested. For that we exposed an additional HTTP endpoint that includes the peer’s ID to get Prometheus metrics for the associated peer. A request to that endpoint makes a gRPC call to the remote node to get the metrics before returning them to the client. With this we achieved the ability to get metrics from all nodes in the drand network, without any additional unneeded traffic.

Reviewer 2: ```suggestion
We also wanted to reduce the amount of metrics traffic between nodes, so we went with a pull system where any node can pull any other node’s metrics when requested. For that we exposed an additional HTTP endpoint that includes the peer’s ID to get Prometheus metrics for the associated peer. A request to that endpoint makes a gRPC call to the remote node to get the metrics before returning them to the client. With this we have the ability to get metrics from all nodes in the drand network, without any unnecessary additional traffic.
```

The Protocol Labs nodes all run instances of [Telegraf](https://www.influxdata.com/time-series-platform/telegraf/), whose configuration is hard-coded to get the metrics from all nodes in the drand network. This is possible because the nodes in the network are static and only change during ceremonies. Telegraf then pushes the metrics into [InfluxDB](https://www.influxdata.com/). We then use [Grafana](https://grafana.com/) to visualize the data.

> Regarding LoE members, the fact that metrics are exposed using the industry-standard Prometheus format and that any node can gather metrics from any other node in the network makes it possible for any of them to create a setup similar to this and gather metrics into their own observability solution.

Reviewer 3: Do we have this open sourced anywhere? perhaps we could link it in case any LoE partners want to copypasta

Reviewer: The Telegraf configuration includes the host/port names of all nodes, so we really don't want to open source it. The way of deploying it is also very specific to the way we run things, I think its usefulness is limited, it's all in the `drand-infra` repo.

Regarding the dashboards themselves, I'm not sure how useful they would be since you would have to be using InfluxDB, my feeling is that most people are using Prometheus for monitoring and that would change the panel definitions completely.

## Dashboard design

> As is usually the case, it took multiple iterations to get to a dashboard design that was useful and also comprehensible. These iterations involved both changing the layout and organization of the presented information, as well as adding metrics that were missed during initial analysis. For example, during ceremonies some of the state changes happened so fast that they were missed by the polling and dashboard refresh intervals, so we added timestamps to be able to visualize whether a state had changed or not. We also reformulated some metrics labels to reduce cardinality.

Reviewer 2: ```suggestion
As is usually the case, it took multiple iterations to get to a dashboard design that was useful and also comprehensible. These iterations involved both changing the layout and organization of the represented information, as well as adding metrics that were missed during initial analysis. For example, during ceremonies some of the state changes are so fast that they were missed by the polling and dashboard refresh intervals, so we added timestamps to be able to visualize whether a state had changed or not. We also reformulated some metrics labels to reduce cardinality.
```

Reviewer 3: disagree with represented suggestion

Reviewer: yeah the information is not represented, it is presented in a dashboard. I do agree with `happened` -> `are`, though.

> We ended up with a dashboard with 2 sections. The first section includes connectivity information for each node, as well as a connectivity matrix showing which nodes are connected to which other nodes. This has enabled us to find problems such as firewall misconfigurations or bugs which impact connectivity.

Reviewer 2: ```suggestion
We ended up with a dashboard with 2 sections. The first section includes connectivity information for each node, as well as a connectivity matrix showing which nodes are connected to which other nodes. This has enabled us to find problems such as firewall misconfigurations or bugs impacting connectivity.
```

![drand_dash_1.png](./images/2022-08-11-observing-randomness/drand_dash_1.png)

> The second section contains information for each of the beacons in the given network, including the ceremony state for each node. This allows us to track ceremonies, as well as ensuring that all nodes are up to date with all other nodes.

Reviewer 2: ```suggestion
The second section contains information about each of the beacons in the given network, including the ceremony state for each node. This allows us to track ceremonies, as well as ensuring that all nodes are up to date with all other nodes.
```

![drand_dash_2.png](./images/2022-08-11-observing-randomness/drand_dash_2.png)

## Results

> We have already deployed and used the new dashboard in the last set of ceremonies that we carried out in our testnet network. It has proven to be an extremely useful tool. We were able to spot connectivity problems and other inconsistencies in a totally async manner, which in turn enabled us to resolve issues much faster. This has given us confidence to run the next testnet ceremony in an asynchronous manner, without having to coordinate all operators to be in front of their keyboard at the same time.

Reviewer 2: ```suggestion
We have already deployed and used the new dashboard in the last set of ceremonies that we carried out in our testnet network. It has proven to be an extremely useful tool. We were able to spot connectivity problems and other inconsistencies in a totally async manner, which in turn enabled us to resolve issues much faster. This has given us confidence to run the next testnet ceremony in an asynchronous manner, without having to coordinate for all operators to be in front of their keyboard at the same time.
```

> We have already deployed and used the new dashboard in the last set of ceremonies that we carried out in our testnet network. It has proven to be an extremely useful tool. We were able to spot connectivity problems and other inconsistencies in a totally async manner, which in turn enabled us to resolve issues much faster. This has given us confidence to run the next testnet ceremony in an asynchronous manner, without having to coordinate for all operators to be in front of their keyboard at the same time.

Reviewer 3: ```suggestion
We have already deployed and used the new dashboard in the last set of ceremonies that we carried out in our testnet network. It has proven to be an extremely useful tool. We were able to spot connectivity problems and other inconsistencies in a totally async manner, which in turn enabled us to resolve issues much faster. This has given us confidence to run the next testnet ceremony in an asynchronous manner, without having to coordinate all operators being in front of their keyboard at the same time.
```

sorry it took me this long to figure out how to use the github suggestion thing 🤦

> Beyond ceremonies, we now have a much more detailed view of the state of the drand network. This has enabled us to debug connectivity problems with some nodes, as well as gain confidence that everything is working correctly.

Reviewer 2: ```suggestion
Beyond ceremonies, we now have a much more detailed view of the state of the LoE drand network.
```
To avoid the repetition, we're already discussion connectivity issues in the previous paragraphs and the confidence boost is mentioned in the next paragraph.
